File: 프로그래머스/3/43164. 여행경로/여행경로.py
from collections import deque

def solution(tickets):
    
    # 그래프 구성
    graph = {}
    for start, end in tickets:
        if start in graph:
            graph[start].append((end))
        else:
            graph[start] = [end]
            
    # 그래프 정렬
    for city in graph:
        graph[city].sort()
        graph[city] = deque(graph[city])
            
    # DFS
    def DFS(start, path):
        nonlocal result, tickets
        
        if len(path) == len(tickets)+1:
            result.append(path)
            return
        
        for idx, cities in enumerate(tickets):
            if cities[0] == start and not visited[idx]:
                visited[idx] = True
                DFS(cities[1], path + [cities[1]])
                visited[idx] = False
    
    result = []
    visited = [False for _ in range(len(tickets))]
    DFS('ICN', ['ICN'])
    result.sort()
    
    return result[0]

File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {
    
    Map<String, List<String>> graph = new HashMap<>();
    int maxMoved = 0;
    
    public String[] solution(String[][] tickets) {
        
        // 그래프 완성
        for (String[] ticket : tickets) {
            String start = ticket[0];
            String end = ticket[1];
            if (!graph.containsKey(start)) graph.put(start, new ArrayList<>());
            graph.get(start).add(end);
        }
        
        // visited, 
        Map<String, boolean[]> visited = new HashMap<>();
        for (String key : graph.keySet()) {
            maxMoved += graph.get(key).size();
            visited.put(key, new boolean[graph.get(key).size()]);
        }
        
        // 정렬
        for (String key : graph.keySet()) {
            graph.get(key).sort((a, b) -> a.compareTo(b));
        }
        
        // dfs
        List<String> answer = new ArrayList<>();
        answer.add("ICN");
        return dfs("ICN", answer, visited, 0).toArray(new String[0]);
    }
    
    public List<String> dfs(String city, List<String> answer, Map<String, boolean[]> visited, int moved) {
        
        // 다 차있으면 끝
        if (moved == maxMoved) {
            return answer;
        }
            
        // 
        List<String> nodes = graph.get(city);
        if (nodes != null) {
            String nextNode;
            for (int i = 0; i < nodes.size(); i++) {
                if (visited.get(city)[i]) continue;
                visited.get(city)[i] = true;
                nextNode = nodes.get(i);
                answer.add(nextNode);
                List<String> result = dfs(nextNode, answer, visited, moved+1);
                if (result != null) return result;
                visited.get(city)[i] = false;
                answer.removeLast();
            }
        }
        
        return null;
    }
    
}